File: Darts/src/code/db/AbstractEntity.java
package code.db;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.sql.Blob;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Pattern;

import code.utils.DatabaseUtil;
import code.utils.SqlErrorConstants;
import object.HandyArrayList;
import util.AbstractClient;
import util.DateUtil;
import util.Debug;
import util.StringUtil;

public abstract class AbstractEntity<E extends AbstractEntity<E>>
					  implements SqlErrorConstants
{
	//statics
	private static final Object UNIQUE_ID_SYNCH_OBJECT = new Object();
	private static ConcurrentHashMap<String, Long> hmLastAssignedIdByTableName = new ConcurrentHashMap<>();
	
	//db fields
	private long rowId = -1;
	private Timestamp dtCreation = DateUtil.getSqlDateNow();
	private Timestamp dtLastUpdate = DateUtil.END_OF_TIME;
	
	//other variables
	private boolean retrievedFromDb = false;
	
	public abstract String getTableName();
	public abstract String getCreateTableSqlSpecific();
	public abstract void populateFromResultSet(E entity, ResultSet rs) throws SQLException;
	public abstract String writeValuesToStatement(PreparedStatement statement, int startIndex, String statementStr)  throws SQLException;
	
	public E factoryFromResultSet(ResultSet rs) throws SQLException
	{
		E ret = factory();
		ret.setRowId(rs.getLong("RowId"));
		ret.setDtCreation(rs.getTimestamp("DtCreation"));
		ret.setDtLastUpdate(rs.getTimestamp("DtLastUpdate"));
		
		populateFromResultSet(ret, rs);
		
		return ret;
	}
	public E factory()
	{
		try
		{
			return (E)getClass().newInstance();
		}
		catch (IllegalAccessException | InstantiationException iae)
		{
			Debug.stackTrace(iae);
			return null;
		}
	}
	
	/**
	 * Default method, override to index your table on creation
	 */
	public void addListsOfColumnsForIndexes(ArrayList<ArrayList<String>> indexes)
	{
		//Do nothing
		indexes.size();
	}
	
	public final boolean columnCanBeUnset(String columnName)
	{
		return getColumnsAllowedToBeUnset().contains(columnName);
	}
	protected ArrayList<String> getColumnsAllowedToBeUnset()
	{
		return new ArrayList<>();
	}
	
	public long assignRowId()
	{
		synchronized (UNIQUE_ID_SYNCH_OBJECT)
		{
			String tableName = getTableName();
			Long lastAssignedIdObj = hmLastAssignedIdByTableName.get(tableName);
			long lastAssignedId = -1;
			if (lastAssignedIdObj == null)
			{
				String query = "SELECT MAX(RowId) FROM " + getTableName();
				
				try (ResultSet rs = DatabaseUtil.executeQuery(query))
				{
					rs.next();
					lastAssignedId = rs.getInt(1);
				}
				catch (SQLException sqle)
				{
					Debug.logSqlException(query, sqle);
				}
			}
			else
			{
				lastAssignedId = lastAssignedIdObj.longValue();
			}
			
			rowId = lastAssignedId + 1;
			hmLastAssignedIdByTableName.put(tableName, rowId);
			
			return rowId;
		}
	}
	
	/**
	 * Alternate DB methods
	 */
	public HandyArrayList<E> retrieveEntitiesAlternateDb(String dbPath)
	{
		return retrieveEntities("", "", dbPath);
	}
 	
	public E retrieveEntity(String whereSql)
	{
		HandyArrayList<E> entities = retrieveEntities(whereSql);
		if (entities.size() > 1)
		{
			Debug.stackTrace("Retrieved " + entities.size() + " rows from " + getTableName() + ". Expected 1. WhereSQL [" + whereSql + "]");
		}
		
		if (entities.isEmpty())
		{
			return null;
		}
		
		return entities.firstElement();
	}
	
	public HandyArrayList<E> retrieveEntities()
	{
		return retrieveEntities("");
	}
	public HandyArrayList<E> retrieveEntities(String whereSql)
	{
		return retrieveEntities(whereSql, "");
	}
	public HandyArrayList<E> retrieveEntities(String whereSql, String alias)
	{
		return retrieveEntities(whereSql, alias, null);
	}
	public HandyArrayList<E> retrieveEntities(String whereSql, String alias, String dbPath)
	{
		String queryWithFrom = "FROM " + getTableName() + " " + alias;
		if (!whereSql.isEmpty())
		{
			queryWithFrom = queryWithFrom + " WHERE " + whereSql;
		}
		
		return retrieveEntitiesWithFrom(queryWithFrom, alias, dbPath);
	}
	public HandyArrayList<E> retrieveEntitiesWithFrom(String whereSqlWithFrom, String alias, String dbPath)
	{
		String query = "SELECT " + getColumnsForSelectStatement(alias) + " " + whereSqlWithFrom;
		
		HandyArrayList<E> ret = new HandyArrayList<>();
		
		try (ResultSet rs = DatabaseUtil.executeQuery(query, dbPath))
		{
			while (rs.next())
			{
				E entity = factoryFromResultSet(rs);
				entity.setRetrievedFromDb(true);
				ret.add(entity);
			}
		}
		catch (SQLException sqle)
		{
			Debug.logSqlException(query, sqle);
		}
		
		return ret;
	}
	
	public E retrieveForId(long rowId)
	{
		return retrieveForId(rowId, true);
	}
	public E retrieveForId(long rowId, boolean stackTraceIfNotFound)
	{
		ArrayList<E> entities = retrieveEntities("RowId = " + rowId);
		if (entities.isEmpty())
		{
			if (stackTraceIfNotFound)
			{
				Debug.stackTrace("Failed to find " + getTableName() + " for ID " + rowId);
			}
			
			return null;
		}
		
		if (entities.size() > 1)
		{
			Debug.stackTrace("Found " + entities.size() + " " + getTableName() + " rows for ID " + rowId);
		}
		
		return entities.get(0);
	}
	
	public boolean deleteFromDatabase()
	{
		String sql = "DELETE FROM " + getTableName() + " WHERE RowId = " + rowId;
		return DatabaseUtil.executeUpdate(sql);
	}
	
	public void saveToDatabase()
	{
		saveToDatabase(DateUtil.getSqlDateNow());
	}
	public void saveToDatabase(Timestamp dtLastUpdate)
	{
		this.dtLastUpdate = dtLastUpdate;
		
		if (retrievedFromDb)
		{
			updateDatabaseRow();
		}
		else
		{
			insertIntoDatabase();
		}
	}
	
	private void updateDatabaseRow()
	{
		String updateQuery = buildUpdateQuery();
		
		try (Connection conn = DatabaseUtil.createDatabaseConnection();
		  PreparedStatement psUpdate = conn.prepareStatement(updateQuery))
		{
			updateQuery = writeTimestamp(psUpdate, 1, getDtCreation(), updateQuery);
			updateQuery = writeTimestamp(psUpdate, 2, getDtLastUpdate(), updateQuery);
			updateQuery = writeValuesToStatement(psUpdate, 3, updateQuery);
			updateQuery = writeLong(psUpdate, getColumnCount(), rowId, updateQuery);
			
			Debug.appendSql(updateQuery, AbstractClient.traceWriteSql);
			
			psUpdate.executeUpdate();
			
			int updateCount = psUpdate.getUpdateCount();
			if (updateCount == 0)
			{
				Debug.stackTrace("0 rows updated: " + updateQuery);
			}
		}
		catch (SQLException sqle)
		{
			Debug.logSqlException(updateQuery, sqle);
		}
	}
	
	private String buildUpdateQuery()
	{
		//Some fun String manipulation
		String columns = getColumnsForSelectStatement();
		columns = columns.replaceAll("RowId, ", "");
		columns = columns.replaceAll(",", "=?,");
		columns += "=?";
		
		String query = "UPDATE " + getTableName()
					 + " SET " + columns
					 + " WHERE RowId=?";
		
		return query;
	}
	
	private void insertIntoDatabase()
	{
		String insertQuery = buildInsertQuery();
		
		try (Connection conn = DatabaseUtil.createDatabaseConnection();
		  PreparedStatement psInsert = conn.prepareStatement(insertQuery);)
		{
			insertQuery = writeLong(psInsert, 1, getRowId(), insertQuery);
			insertQuery = writeTimestamp(psInsert, 2, getDtCreation(), insertQuery);
			insertQuery = writeTimestamp(psInsert, 3, getDtLastUpdate(), insertQuery);
			insertQuery = writeValuesToStatement(psInsert, 4, insertQuery);
			
			Debug.appendSql(insertQuery, AbstractClient.traceWriteSql);
			
			psInsert.executeUpdate();
			
			//Set this so we can call save() again on the same java object and get the right behaviour
			retrievedFromDb = true;
		}
		catch (SQLException sqle)
		{
			Debug.logSqlException(insertQuery, sqle);
		}
	}
	
	private String buildInsertQuery()
	{
		StringBuilder sbInsert = new StringBuilder();
		sbInsert.append("INSERT INTO ");
		sbInsert.append(getTableName());
		sbInsert.append(" VALUES (");
		
		for (int i=0; i<getColumnCount(); i++)
		{
			if (i > 0)
			{
				sbInsert.append(", ");
			}
			
			sbInsert.append("?");
		}
		
		sbInsert.append(")");
		return sbInsert.toString();
	}
	
	private int getColumnCount()
	{
		String columns = getColumnsForSelectStatement();
		return StringUtil.countOccurences(columns, ",") + 1;
	}
	
	public boolean createTable()
	{
		boolean createdTable = DatabaseUtil.createTableIfNotExists(getTableName(), getCreateTableColumnSql());
		if (createdTable)
		{
			createIndexes();
		}
		
		return createdTable;
	}
	public void createIndexes()
	{
		//Also create the indexes
		ArrayList<ArrayList<String>> indexes = new ArrayList<>();
		addListsOfColumnsForIndexes(indexes);
		
		for (int i=0; i<indexes.size(); i++)
		{
			ArrayList<String> columnsForIndex = indexes.get(i);
			createIndex(columnsForIndex);
		}
	}
	private void createIndex(ArrayList<String> columns)
	{
		String columnList = StringUtil.toDelims(columns, ",");
		String indexName = columnList.replaceAll(",", "_");
		
		String statement = "CREATE INDEX " + indexName + " ON " + getTableName() + "(" + columnList + ")";
		boolean success = DatabaseUtil.executeUpdate(statement);
		if (!success)
		{
			Debug.append("Failed to create index " + indexName + " on " + getTableName());
		}
	}
	
	public boolean addIntColumn(String columnName)
	{
		if (columnExists(columnName))
		{
			Debug.append("Not adding column " + columnName + " to " + getTableName() + " as it already exists");
			return false;
		}
		
		String sql = "ALTER TABLE " + getTableName() + " ADD COLUMN " + columnName + " INT NOT NULL DEFAULT -1";
		boolean addedColumn = DatabaseUtil.executeUpdate(sql);
		if (!addedColumn)
		{
			return false;
		}
		
		//We've added the column, now attempt to drop the default.
		String defaultSql = "ALTER TABLE " + getTableName() + " ALTER COLUMN " + columnName + " DEFAULT NULL";
		return DatabaseUtil.executeUpdate(defaultSql);
	}
	private boolean columnExists(String columnName)
	{
		columnName = columnName.toUpperCase();
		String tableName = getTableName().toUpperCase();
		
		StringBuilder sb = new StringBuilder();
		sb.append("SELECT COUNT(1) ");
		sb.append("FROM sys.systables t, sys.syscolumns c ");
		sb.append("WHERE c.ReferenceId = t.TableId ");
		sb.append("AND t.TableName = '");
		sb.append(tableName);
		sb.append("' AND c.ColumnName = '");
		sb.append(columnName);
		sb.append("'");
		
		int count = DatabaseUtil.executeQueryAggregate(sb);
		return count > 0;
	}
	
	private String getCreateTableColumnSql()
	{
		return "RowId int PRIMARY KEY, DtCreation Timestamp NOT NULL, DtLastUpdate Timestamp NOT NULL, " + getCreateTableSqlSpecific();
	}
	
	public final ArrayList<String> getColumns()
	{
		ArrayList<String> columns = new ArrayList<>();
		
		String columnCreateSql = getCreateTableColumnSql();
		
		ArrayList<String> cols = StringUtil.getListFromDelims(columnCreateSql, ",");
		for (int i=0; i<cols.size(); i++)
		{
			String rawColumn = cols.get(i);
			String cleanColumn = getColumnNameFromCreateSql(rawColumn);
			
			columns.add(cleanColumn);
		}
		
		return columns;
	}
	
	private String getColumnsForSelectStatement()
	{
		return getColumnsForSelectStatement("");
	}
	private String getColumnsForSelectStatement(String alias)
	{
		StringBuilder sb = new StringBuilder();
		
		ArrayList<String> cols = getColumns();
		for (int i=0; i<cols.size(); i++)
		{
			if (i > 0)
			{
				sb.append(", ");
			}
			
			String column = cols.get(i);
			if (!alias.isEmpty())
			{
				column = alias + "." + column;
			}
			
			sb.append(column);
		}
		
		return sb.toString();
	}
	private String getColumnNameFromCreateSql(String col)
	{
		col = col.trim();
		col = col.replace("(", "");
		col = col.replace(")", "");
		
		int spaceIndex = col.indexOf(' ');
		col = col.substring(0, spaceIndex);
		
		return col;
	}
	
	/**
	 * Write to statement methods
	 */
	public String writeLong(PreparedStatement ps, int ix, long value, String statementStr) throws SQLException
	{
		ps.setLong(ix, value);
		return swapInValue(statementStr, value);
	}
	public String writeInt(PreparedStatement ps, int ix, int value, String statementStr) throws SQLException
	{
		ps.setInt(ix, value);
		return swapInValue(statementStr, value);
	}
	public String writeString(PreparedStatement ps, int ix, String value, String statementStr) throws SQLException
	{
		ps.setString(ix, value);
		return swapInValue(statementStr, value);
	}
	public String writeTimestamp(PreparedStatement ps, int ix, Timestamp value, String statementStr) throws SQLException
	{
		ps.setTimestamp(ix, value);
		return swapInValue(statementStr, value);
	}
	public String writeBlob(PreparedStatement ps, int ix, Blob value, String statementStr) throws SQLException
	{
		ps.setBlob(ix, value);
		String blobStr = "Blob (dataLength: " + value.length() + ")";
		return swapInValue(statementStr, blobStr);
	}
	public String writeBoolean(PreparedStatement ps, int ix, boolean value, String statementStr) throws SQLException
	{
		ps.setBoolean(ix, value);
		return swapInValue(statementStr, value);
	}
	private String swapInValue(String statementStr, Object value)
	{
		statementStr = statementStr.replaceFirst(Pattern.quote("?"), "" + value);
		return statementStr;
	}
	
	public static <E> HandyArrayList<E> makeFromEntityFields(ArrayList<AbstractEntity<?>> entities, String fieldName)
	{
		HandyArrayList<E> ret = new HandyArrayList<>();
		if (entities.isEmpty())
		{
			return ret;
		}
		
		try
		{
			String getMethod = "get" + fieldName;
			
			for (AbstractEntity<?> entity : entities)
			{
				Class<AbstractEntity<?>> c = (Class<AbstractEntity<?>>)entity.getClass();
				Method m = c.getMethod(getMethod, new Class[0]);
				E obj = (E)m.invoke(entity, new Object[0]);
				ret.add(obj);
			}
		}
		catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e)
		{
			Debug.stackTrace(e, "Reflection error making field list [" + fieldName + "] for entities: " + entities);
		}
		
		return ret;	
		
		
	}
	
	public String getTableNameUpperCase()
	{
		return getTableName().toUpperCase();
	}
	
	/**
	 * Gets / Sets
	 */
	public long getRowId()
	{
		return rowId;
	}
	public void setRowId(long rowId)
	{
		this.rowId = rowId;
	}
	public Timestamp getDtCreation()
	{
		return dtCreation;
	}
	public void setDtCreation(Timestamp dtCreation)
	{
		this.dtCreation = dtCreation;
	}
	public Timestamp getDtLastUpdate()
	{
		return dtLastUpdate;
	}
	public void setDtLastUpdate(Timestamp dtLastUpdate)
	{
		this.dtLastUpdate = dtLastUpdate;
	}
	public boolean getRetrievedFromDb()
	{
		return retrievedFromDb;
	}
	public void setRetrievedFromDb(boolean retrievedFromDb)
	{
		this.retrievedFromDb = retrievedFromDb;
	}
}

File: src/main/kotlin/dartzee/db/AbstractEntity.kt
package dartzee.db

import dartzee.`object`.SegmentType
import dartzee.achievements.AchievementType
import dartzee.core.util.DateStatics
import dartzee.core.util.getSqlDateNow
import dartzee.core.util.getSqlString
import dartzee.game.GameType
import dartzee.game.MatchMode
import dartzee.logging.*
import dartzee.logging.exceptions.ApplicationFault
import dartzee.logging.exceptions.WrappedSqlException
import dartzee.utils.Database
import dartzee.utils.DurationTimer
import dartzee.utils.InjectedThings
import dartzee.utils.InjectedThings.logger
import java.sql.*
import java.util.*
import java.util.regex.Pattern

abstract class AbstractEntity<E : AbstractEntity<E>>(protected val database: Database = InjectedThings.mainDatabase)
{
    //DB Fields
    var rowId: String = ""
    var dtCreation = getSqlDateNow()
    var dtLastUpdate = DateStatics.END_OF_TIME

    //other variables
    var retrievedFromDb = false

    /**
     * Abstract fns
     */
    abstract fun getTableName(): EntityName
    abstract fun getCreateTableSqlSpecific(): String

    /**
     * Default implementations
     */
    open fun getColumnsAllowedToBeUnset() = listOf<String>()
    open fun addListsOfColumnsForIndexes(indexes: MutableList<List<String>>) {}
    open fun cacheValuesWhileResultSetActive() {}
    open fun includeInSync() = true

    /**
     * Helpers
     */
    private fun getColumnCount() = getColumns().size

    private fun getCreateTableColumnSql() = "RowId VARCHAR(36) PRIMARY KEY, DtCreation Timestamp NOT NULL, DtLastUpdate Timestamp NOT NULL, ${getCreateTableSqlSpecific()}"

    fun getColumns(): MutableList<String>
    {
        val columnCreateSql = getCreateTableColumnSql()
        val cols = columnCreateSql.split(",")

        return cols.map{ getColumnNameFromCreateSql(it) }.toMutableList()
    }

    private fun getColumnsExcluding(vararg columnsToExclude: String): MutableList<String>
    {
        val columns = getColumns()
        columnsToExclude.forEach { columns.remove(it) }
        return columns
    }

    fun getTableNameUpperCase() = getTableName().name.uppercase()

    fun factoryFromResultSet(rs: ResultSet): E
    {
        val ret = factory()!!

        ret.rowId = rs.getString("RowId")
        ret.dtCreation = rs.getTimestamp("DtCreation")
        ret.dtLastUpdate = rs.getTimestamp("DtLastUpdate")
        ret.retrievedFromDb = true

        getColumnsExcluding("RowId", "DtCreation", "DtLastUpdate").forEach{
            val rsValue = getFieldFromResultSet(rs, it)
            ret.setField(it, rsValue)
        }

        ret.cacheValuesWhileResultSetActive()

        return ret
    }

    @Suppress("UNCHECKED_CAST")
    private fun factory(db: Database = database): E? =
        try
        {
            javaClass.getDeclaredConstructor(Database::class.java).newInstance(db) as E
        }
        catch (t: Throwable)
        {
            logger.error(CODE_INSTANTIATION_ERROR, "Failed to instantiate ${javaClass.simpleName}", t)
            null
        }


    fun columnCanBeUnset(columnName: String) = getColumnsAllowedToBeUnset().contains(columnName)

    open fun assignRowId(): String
    {
        rowId = UUID.randomUUID().toString()
        return rowId
    }

    fun retrieveEntity(whereSql: String): E?
    {
        val entities = retrieveEntities(whereSql)
        if (entities.size > 1)
        {
            logger.error(CODE_SQL_EXCEPTION,
                    "Retrieved ${entities.size} rows from ${getTableName()}. Expected 1. WhereSQL [$whereSql]",
                    Throwable(),
                    KEY_SQL to whereSql)
        }

        return entities.firstOrNull()
    }

    fun retrieveEntities(whereSql: String = "", alias: String = ""): MutableList<E>
    {
        var queryWithFrom = "FROM ${getTableName()}"
        if (alias.isNotEmpty())
        {
            queryWithFrom += " $alias"
        }

        if (whereSql.isNotEmpty())
        {
            queryWithFrom += " WHERE $whereSql"
        }

        return retrieveEntitiesWithFrom(queryWithFrom, alias)
    }

    private fun retrieveEntitiesWithFrom(whereSqlWithFrom: String, alias: String): MutableList<E>
    {
        val query = "SELECT " + getColumnsForSelectStatement(alias) + " " + whereSqlWithFrom
        val ret = mutableListOf<E>()

        database.executeQuery(query).use { rs ->
            while (rs.next())
            {
                val entity = factoryFromResultSet(rs)
                ret.add(entity)
            }
        }

        return ret
    }

    fun retrieveForId(rowId: String, stackTraceIfNotFound: Boolean = true): E?
    {
        val entity = retrieveEntity("RowId = '$rowId'")
        if (entity == null && stackTraceIfNotFound)
        {
            logger.error(CODE_SQL_EXCEPTION,
                    "Failed to find ${getTableName()} for ID [$rowId]",
                    Throwable(),
                    KEY_SQL to "RowId = '$rowId'")
        }


        return entity
    }

    fun deleteFromDatabase(): Boolean
    {
        val sql = "DELETE FROM ${getTableName()} WHERE RowId = '$rowId'"
        val success = database.executeUpdate(sql)
        if (success && includeInSync()) {
            DeletionAuditEntity.factoryAndSave(this)
        }

        return success
    }

    fun deleteAll()
    {
        if (includeInSync())
        {
            logger.error(CODE_DELETE_ERROR, "Wiping of ${getTableName()} will not be audited. This will break the sync!")
        }

        database.executeUpdate("DELETE FROM ${getTableName()}")
    }

    fun deleteWhere(whereSql: String): Boolean
    {
        val audits = retrieveEntities(whereSql).map { DeletionAuditEntity.factory(it) }
        val sql = "DELETE FROM ${getTableName()} WHERE $whereSql"
        val success = database.executeUpdate(sql)
        if (success && includeInSync()) {
            BulkInserter.insert(audits)
        }

        return success
    }

    fun saveToDatabase(dtLastUpdate: Timestamp = getSqlDateNow())
    {
        this.dtLastUpdate = dtLastUpdate

        if (retrievedFromDb)
        {
            updateDatabaseRow()
        }
        else
        {
            insertIntoDatabase()
        }
    }

    fun count() = countWhere("")
    fun countWhere(whereSql: String): Int
    {
        val fullWhere = if (whereSql.isNotEmpty()) "WHERE $whereSql" else ""
        return database.executeQueryAggregate("SELECT COUNT(1) FROM ${getTableName()} $fullWhere")
    }

    /**
     * Merge helpers
     */
    fun countModifiedSince(dt: Timestamp?): Int
    {
        val whereSql = if (dt != null) "DtLastUpdate > ${dt.getSqlString()}" else ""
        return countWhere(whereSql)
    }
    fun retrieveModifiedSince(dt: Timestamp?): List<E>
    {
        val whereSql = if (dt != null) "DtLastUpdate > ${dt.getSqlString()}" else ""
        return retrieveEntities(whereSql).sortedBy { it.dtCreation }
    }

    fun mergeIntoDatabase(otherDatabase: Database)
    {
        val otherDao = factory(otherDatabase) ?: throw ApplicationFault(CODE_MERGE_ERROR, "Failed to factory ${getTableName()} dao")
        val existingRow = otherDao.retrieveForId(rowId, false)
        if (existingRow == null)
        {
            reassignLocalId(otherDatabase)
            insertIntoDatabase(otherDatabase)
        }
        else if (dtLastUpdate.after(existingRow.dtLastUpdate))
        {
            updateDatabaseRow(otherDatabase)
        }

        mergeImpl(otherDatabase)
    }
    open fun reassignLocalId(otherDatabase: Database) {}
    open fun mergeImpl(otherDatabase: Database) {}

    private fun updateDatabaseRow(db: Database = database)
    {
        val genericUpdate = buildUpdateQuery()
        var updateQuery = genericUpdate

        val conn = db.borrowConnection()
        try
        {
            conn.prepareStatement(updateQuery).use { psUpdate ->
                updateQuery = writeValuesToStatement(psUpdate, 1, updateQuery)
                updateQuery = writeString(psUpdate, getColumnCount(), rowId, updateQuery)

                val timer = DurationTimer()
                psUpdate.executeUpdate()

                val updateCount = psUpdate.updateCount

                logger.logSql(updateQuery, genericUpdate, timer.getDuration(), updateCount, db.dbName)

                if (updateCount == 0)
                {
                    logger.error(CODE_SQL_EXCEPTION,
                            "0 rows updated for statement $updateQuery",
                            Throwable(),
                            KEY_SQL to updateQuery)
                }
            }
        }
        catch (sqle: SQLException)
        {
            throw WrappedSqlException(updateQuery, genericUpdate, sqle)
        }
        finally
        {
            db.returnConnection(conn)
        }
    }

    private fun buildUpdateQuery(): String
    {
        //Some fun String manipulation
        var columns = getColumnsForSelectStatement()
        columns = columns.replaceFirst("RowId, ", "")
        columns = columns.replace(",", "=?,")
        columns += "=?"

        return "UPDATE ${getTableName()} SET $columns WHERE RowId=?"
    }

    private fun insertIntoDatabase(db: Database = database)
    {
        val genericInsert = "INSERT INTO ${getTableName()} VALUES ${getInsertBlockForStatement()}"
        var insertQuery = genericInsert

        val conn = db.borrowConnection()
        try
        {
            conn.prepareStatement(insertQuery).use { psInsert ->
                insertQuery = writeValuesToInsertStatement(insertQuery, psInsert)

                val timer = DurationTimer()
                psInsert.executeUpdate()
                logger.logSql(insertQuery, genericInsert, timer.getDuration(), psInsert.updateCount, db.dbName)

                //Set this so we can call save() again on the same object and get the right behaviour
                retrievedFromDb = true
            }
        }
        catch (sqle: SQLException)
        {
            throw WrappedSqlException(insertQuery, genericInsert, sqle)
        }
        finally
        {
            db.returnConnection(conn)
        }
    }

    fun writeValuesToInsertStatement(emptyStatement: String, psInsert: PreparedStatement, entityNumber: Int = 0): String
    {
        val adjustment = entityNumber * getColumnCount()

        var insertQuery = emptyStatement
        insertQuery = writeString(psInsert, 1 + adjustment, rowId, insertQuery)
        insertQuery = writeValuesToStatement(psInsert, 2 + adjustment, insertQuery)

        return insertQuery
    }

    private fun writeValuesToStatement(ps: PreparedStatement, startIx: Int, emptyStatement: String): String
    {
        var ix = startIx
        var statementStr = emptyStatement
        getColumnsExcluding("RowId").forEach {
            statementStr = writeValue(ps, ix++, it, statementStr)
        }

        return statementStr
    }

    fun getInsertBlockForStatement() = "(${getColumns().joinToString{"?"}})"

    open fun createTable(): Boolean
    {
        val createdTable = database.createTableIfNotExists(getTableName().name, getCreateTableColumnSql())
        if (createdTable)
        {
            createIndexes()
        }

        return createdTable
    }

    private fun createIndexes()
    {
        //Also create the indexes
        val indexes = mutableListOf<List<String>>()
        addListsOfColumnsForIndexes(indexes)

        indexes.forEach{
            createIndex(it)
        }
    }

    private fun createIndex(columns: List<String>)
    {
        val columnList = columns.joinToString()
        val indexName = columnList.replace(", ", "_")

        val statement = "CREATE INDEX $indexName ON ${getTableName()}($columnList)"
        database.executeUpdate(statement)
    }

    fun getColumnsForSelectStatement(alias: String = ""): String
    {
        var cols = getColumns().toList()
        if (alias.isNotEmpty())
        {
            cols = cols.map{ "$alias.$it" }
        }

        return cols.joinToString()
    }

    private fun getColumnNameFromCreateSql(col: String): String
    {
        var colSanitised = col
        colSanitised = colSanitised.trim()
        colSanitised = colSanitised.replace("(", "")
        colSanitised = colSanitised.replace(")", "")

        return colSanitised.split(" ")[0]
    }

    fun getField(fieldName: String): Any?
    {
        val getter = javaClass.getMethod("get$fieldName")
        return getter.invoke(this)
    }
    fun setField(fieldName: String, value: Any?)
    {
        val getMethod = javaClass.getMethod("get$fieldName")
        val setMethod = javaClass.getDeclaredMethod("set$fieldName", getMethod.returnType)

        setMethod.invoke(this, value)
    }
    fun getFieldType(fieldName: String): Class<*>
    {
        val getter = javaClass.getMethod("get$fieldName")
        return getter.returnType
    }

    /**
     * Write to statement methods
     */
    private fun writeLong(ps: PreparedStatement, ix: Int, value: Long, statementStr: String): String
    {
        ps.setLong(ix, value)
        return swapInValue(statementStr, value)
    }

    private fun writeInt(ps: PreparedStatement, ix: Int, value: Int, statementStr: String): String
    {
        ps.setInt(ix, value)
        return swapInValue(statementStr, value)
    }

    private fun writeDouble(ps: PreparedStatement, ix: Int, value: Double, statementStr: String): String
    {
        ps.setDouble(ix, value)
        return swapInValue(statementStr, value)
    }

    private fun writeString(ps: PreparedStatement, ix: Int, value: String, statementStr: String): String
    {
        ps.setString(ix, value)
        return swapInValue(statementStr, "'$value'")
    }

    private fun writeTimestamp(ps: PreparedStatement, ix: Int, value: Timestamp, statementStr: String): String
    {
        ps.setTimestamp(ix, value)
        return swapInValue(statementStr, "'$value'")
    }

    private fun writeBlob(ps: PreparedStatement, ix: Int, value: Blob, statementStr: String): String
    {
        ps.setBlob(ix, value)
        val blobStr = "Blob (dataLength: " + value.length() + ")"
        return swapInValue(statementStr, blobStr)
    }

    private fun writeBoolean(ps: PreparedStatement, ix: Int, value: Boolean, statementStr: String): String
    {
        ps.setBoolean(ix, value)
        return swapInValue(statementStr, value)
    }

    private fun swapInValue(statementStr: String, value: Any) = statementStr.replaceFirst(Pattern.quote("?").toRegex(), "" + value)

    private fun writeValue(ps: PreparedStatement, ix: Int, columnName: String, statementStr: String): String
    {
        val value = getField(columnName)
        return when (getFieldType(columnName))
        {
            String::class.java -> writeString(ps, ix, value as String, statementStr)
            Long::class.java -> writeLong(ps, ix, value as Long, statementStr)
            Int::class.java -> writeInt(ps, ix, value as Int, statementStr)
            Boolean::class.java -> writeBoolean(ps, ix, value as Boolean, statementStr)
            Timestamp::class.java -> writeTimestamp(ps, ix, value as Timestamp, statementStr)
            Blob::class.java -> writeBlob(ps, ix, value as Blob, statementStr)
            Double::class.java -> writeDouble(ps, ix, value as Double, statementStr)
            else -> writeString(ps, ix, "$value", statementStr)
        }
    }

    private fun getFieldFromResultSet(rs: ResultSet, columnName: String): Any? =
        when(getFieldType(columnName))
        {
            String::class.java -> rs.getString(columnName)
            Long::class.java -> rs.getLong(columnName)
            Int::class.java -> rs.getInt(columnName)
            Boolean::class.java -> rs.getBoolean(columnName)
            Timestamp::class.java -> rs.getTimestamp(columnName)
            Blob::class.java -> rs.getBlob(columnName)
            Double::class.java -> rs.getDouble(columnName)
            GameType::class.java -> GameType.valueOf(rs.getString(columnName))
            MatchMode::class.java -> MatchMode.valueOf(rs.getString(columnName))
            SegmentType::class.java -> SegmentType.valueOf(rs.getString(columnName))
            AchievementType::class.java -> AchievementType.valueOf(rs.getString(columnName))
            EntityName::class.java -> EntityName.valueOf(rs.getString(columnName))
            else -> null
        }

    private fun getValueForLogging(value: Any) =
        when (value.javaClass)
        {
            String::class.java, Timestamp::class.java -> "'$value'"
            Blob::class.java -> "BLOB:${(value as Blob).length()}"
            else -> "$value"
        }
}
